File: fase4/pillar-c/otel-sdk/go/metrics.go
package main

import (
	"context"
	"log"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/metric"
)

// MantaMetrics holds all OpenTelemetry metrics for Manta observability.
type MantaMetrics struct {
	// Git metrics
	GitMergeSuccessCount     metric.Int64Counter
	GitMergeFailureCount     metric.Int64Counter
	GitMergeSuccessRate      metric.Float64Gauge
	GitPRReviewTime          metric.Float64Histogram
	GitConflictResolutionTime metric.Float64Histogram
	GitCommitSize            metric.Int64Histogram
	GitCIDuration            metric.Float64Histogram

	// ML metrics
	MLModelAccuracy          metric.Float64Gauge
	MLModelAccuracyBaseline  metric.Float64Gauge
	MLModelPrecision         metric.Float64Gauge
	MLModelRecall            metric.Float64Gauge
	MLInferenceLatency       metric.Float64Histogram
	MLTrainingDuration       metric.Float64Histogram
	MLFeatureImportance      metric.Float64Gauge
	MLPredictionCount        metric.Int64Counter

	// Infrastructure metrics
	CPUUsagePercent          metric.Float64Gauge
	MemoryUsageMB            metric.Int64Gauge
	DiskIOOperations         metric.Int64Counter
	NetworkThroughputMBps    metric.Float64Gauge
	PodRestartCount          metric.Int64Counter

	// Business metrics
	CostPerMergeDollars      metric.Float64Gauge
	ROIPerFeature            metric.Float64Gauge
	VelocityMergesPerDay     metric.Float64Gauge
	CostMonthly              metric.Float64Gauge
	FeatureDeploymentTime    metric.Float64Histogram

	// Anomaly metrics
	AnomalyDetectionLatency  metric.Float64Histogram
	AnomalyScore             metric.Float64Gauge
	PatternQualityScore      metric.Float64Gauge
	CanaryRolloutProgress    metric.Float64Gauge
	AnomalyUnresolvedCount   metric.Int64Gauge
	DBSCANClusterSize        metric.Int64Gauge
	CanaryPhaseChangeTime    metric.Int64Gauge
	JaegerSpanErrorRate      metric.Float64Gauge
}

// InitializeMetrics sets up all Manta metrics.
func InitializeMetrics(ctx context.Context) (*MantaMetrics, error) {
	meter := otel.Meter("manta-observability")
	m := &MantaMetrics{}

	// Git metrics
	var err error
	m.GitMergeSuccessCount, err = meter.Int64Counter(
		"manta_git_merge_success_total",
		metric.WithDescription("Total number of successful PR merges"),
		metric.WithUnit("1"),
	)
	if err != nil {
		log.Printf("Failed to create GitMergeSuccessCount: %v\n", err)
		return nil, err
	}

	m.GitMergeFailureCount, err = meter.Int64Counter(
		"manta_git_merge_failure_total",
		metric.WithDescription("Total number of failed PR merges"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	m.GitMergeSuccessRate, err = meter.Float64Gauge(
		"manta_git_merge_success_rate",
		metric.WithDescription("PR merge success rate (0-1)"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	m.GitPRReviewTime, err = meter.Float64Histogram(
		"manta_git_pr_review_time_seconds",
		metric.WithDescription("Time taken to review a PR from creation to merge"),
		metric.WithUnit("s"),
	)
	if err != nil {
		return nil, err
	}

	m.GitConflictResolutionTime, err = meter.Float64Histogram(
		"manta_git_conflict_resolution_time_seconds",
		metric.WithDescription("Time to resolve merge conflicts"),
		metric.WithUnit("s"),
	)
	if err != nil {
		return nil, err
	}

	m.GitCommitSize, err = meter.Int64Histogram(
		"manta_git_commit_size_lines",
		metric.WithDescription("Number of lines changed per commit"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	m.GitCIDuration, err = meter.Float64Histogram(
		"manta_git_ci_duration_seconds",
		metric.WithDescription("CI/CD pipeline execution duration"),
		metric.WithUnit("s"),
	)
	if err != nil {
		return nil, err
	}

	// ML metrics
	m.MLModelAccuracy, err = meter.Float64Gauge(
		"manta_ml_model_accuracy",
		metric.WithDescription("Current ML model accuracy (0-1)"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	m.MLInferenceLatency, err = meter.Float64Histogram(
		"manta_ml_inference_latency_ms",
		metric.WithDescription("ML model inference latency"),
		metric.WithUnit("ms"),
	)
	if err != nil {
		return nil, err
	}

	m.MLPredictionCount, err = meter.Int64Counter(
		"manta_ml_predictions_total",
		metric.WithDescription("Total number of ML model predictions"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	// Infrastructure metrics
	m.CPUUsagePercent, err = meter.Float64Gauge(
		"manta_infrastructure_cpu_usage_percent",
		metric.WithDescription("CPU usage percentage (0-100)"),
		metric.WithUnit("%"),
	)
	if err != nil {
		return nil, err
	}

	m.MemoryUsageMB, err = meter.Int64Gauge(
		"manta_infrastructure_memory_usage_mb",
		metric.WithDescription("Memory usage in megabytes"),
		metric.WithUnit("MB"),
	)
	if err != nil {
		return nil, err
	}

	// Business metrics
	m.CostPerMergeDollars, err = meter.Float64Gauge(
		"manta_cost_per_merge_dollars",
		metric.WithDescription("Infrastructure cost per successful merge"),
		metric.WithUnit("$"),
	)
	if err != nil {
		return nil, err
	}

	m.CostMonthly, err = meter.Float64Gauge(
		"manta_cost_monthly",
		metric.WithDescription("Monthly infrastructure cost"),
		metric.WithUnit("$"),
	)
	if err != nil {
		return nil, err
	}

	// Anomaly metrics
	m.AnomalyDetectionLatency, err = meter.Float64Histogram(
		"manta_anomaly_detection_latency_ms",
		metric.WithDescription("Latency of anomaly detection inference"),
		metric.WithUnit("ms"),
	)
	if err != nil {
		return nil, err
	}

	m.AnomalyScore, err = meter.Float64Gauge(
		"manta_anomaly_score",
		metric.WithDescription("Raw anomaly score from Isolation Forest (0-1)"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	m.AnomalyUnresolvedCount, err = meter.Int64Gauge(
		"manta_anomaly_detection_unresolved_count",
		metric.WithDescription("Number of unresolved anomalies"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	m.DBSCANClusterSize, err = meter.Int64Gauge(
		"manta_anomaly_dbscan_cluster_size",
		metric.WithDescription("Size of largest DBSCAN cluster (indicates drift magnitude)"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	m.JaegerSpanErrorRate, err = meter.Float64Gauge(
		"manta_jaeger_span_error_rate",
		metric.WithDescription("Jaeger span error rate (0-1)"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	log.Println("Manta metrics initialized successfully")
	return m, nil
}

File: fase4/pillar-c/otel-sdk/python/metrics.py
"""
Prometheus metrics definitions for Manta Maestro observability.
Covers git, ML, infrastructure, business, and anomaly detection domains.
"""

from opentelemetry import metrics
from opentelemetry.sdk.metrics.aggregation import (
    ExplicitBucketHistogramAggregation,
    LastValueAggregation,
    SumAggregation,
)
from typing import Optional


class MantaMetrics:
    """Centralized metrics management for Manta observability."""

    def __init__(self):
        self.meter = metrics.get_meter("manta-observability")
        self._setup_git_metrics()
        self._setup_ml_metrics()
        self._setup_infrastructure_metrics()
        self._setup_business_metrics()
        self._setup_anomaly_metrics()

    def _setup_git_metrics(self) -> None:
        """Setup Git/GitOps metrics."""
        # Git merge metrics
        self.git_merge_success_count = self.meter.create_counter(
            name="manta_git_merge_success_total",
            description="Total number of successful PR merges",
            unit="1",
        )

        self.git_merge_failure_count = self.meter.create_counter(
            name="manta_git_merge_failure_total",
            description="Total number of failed PR merges",
            unit="1",
        )

        self.git_merge_success_rate = self.meter.create_gauge(
            name="manta_git_merge_success_rate",
            description="PR merge success rate (0-1)",
            unit="1",
        )

        self.git_pr_review_time = self.meter.create_histogram(
            name="manta_git_pr_review_time_seconds",
            description="Time taken to review a PR from creation to merge",
            unit="s",
            aggregation=ExplicitBucketHistogramAggregation([60, 300, 900, 1800, 3600, 7200]),
        )

        self.git_conflict_resolution_time = self.meter.create_histogram(
            name="manta_git_conflict_resolution_time_seconds",
            description="Time to resolve merge conflicts",
            unit="s",
            aggregation=ExplicitBucketHistogramAggregation([60, 300, 900, 1800, 3600, 7200]),
        )

        self.git_commit_size = self.meter.create_histogram(
            name="manta_git_commit_size_lines",
            description="Number of lines changed per commit",
            unit="1",
            aggregation=ExplicitBucketHistogramAggregation([10, 50, 100, 500, 1000, 5000]),
        )

        self.git_ci_duration = self.meter.create_histogram(
            name="manta_git_ci_duration_seconds",
            description="CI/CD pipeline execution duration",
            unit="s",
            aggregation=ExplicitBucketHistogramAggregation([30, 60, 120, 300, 600, 1200]),
        )

    def _setup_ml_metrics(self) -> None:
        """Setup ML/AI metrics."""
        self.ml_model_accuracy = self.meter.create_gauge(
            name="manta_ml_model_accuracy",
            description="Current ML model accuracy (0-1)",
            unit="1",
        )

        self.ml_model_accuracy_baseline = self.meter.create_gauge(
            name="manta_ml_model_accuracy_baseline",
            description="Baseline ML model accuracy for drift detection",
            unit="1",
        )

        self.ml_model_precision = self.meter.create_gauge(
            name="manta_ml_model_precision",
            description="ML model precision metric",
            unit="1",
        )

        self.ml_model_recall = self.meter.create_gauge(
            name="manta_ml_model_recall",
            description="ML model recall metric",
            unit="1",
        )

        self.ml_inference_latency = self.meter.create_histogram(
            name="manta_ml_inference_latency_ms",
            description="ML model inference latency",
            unit="ms",
            aggregation=ExplicitBucketHistogramAggregation([10, 50, 100, 500, 1000, 2000]),
        )

        self.ml_training_duration = self.meter.create_histogram(
            name="manta_ml_training_duration_hours",
            description="ML model training duration",
            unit="h",
            aggregation=ExplicitBucketHistogramAggregation([0.5, 1, 2, 4, 8, 24]),
        )

        self.ml_feature_importance = self.meter.create_gauge(
            name="manta_ml_feature_importance",
            description="Importance score of individual features (top 10)",
            unit="1",
        )

        self.ml_prediction_count = self.meter.create_counter(
            name="manta_ml_predictions_total",
            description="Total number of ML model predictions",
            unit="1",
        )

    def _setup_infrastructure_metrics(self) -> None:
        """Setup infrastructure metrics."""
        self.cpu_usage_percent = self.meter.create_gauge(
            name="manta_infrastructure_cpu_usage_percent",
            description="CPU usage percentage (0-100)",
            unit="%",
        )

        self.memory_usage_mb = self.meter.create_gauge(
            name="manta_infrastructure_memory_usage_mb",
            description="Memory usage in megabytes",
            unit="MB",
        )

        self.disk_io_operations = self.meter.create_counter(
            name="manta_infrastructure_disk_io_operations_total",
            description="Total disk I/O operations",
            unit="1",
        )

        self.network_throughput_mbps = self.meter.create_gauge(
            name="manta_infrastructure_network_throughput_mbps",
            description="Network throughput in Mbps",
            unit="Mbps",
        )

        self.pod_restart_count = self.meter.create_counter(
            name="manta_infrastructure_pod_restarts_total",
            description="Total number of pod restarts",
            unit="1",
        )

    def _setup_business_metrics(self) -> None:
        """Setup business/cost metrics."""
        self.cost_per_merge_dollars = self.meter.create_gauge(
            name="manta_cost_per_merge_dollars",
            description="Infrastructure cost per successful merge",
            unit="$",
        )

        self.roi_per_feature = self.meter.create_gauge(
            name="manta_roi_per_feature",
            description="Return on investment per feature deployed",
            unit="1",
        )

        self.velocity_merges_per_day = self.meter.create_gauge(
            name="manta_velocity_merges_per_day",
            description="Average merges per day (deployment velocity)",
            unit="1",
        )

        self.cost_monthly = self.meter.create_gauge(
            name="manta_cost_monthly",
            description="Monthly infrastructure cost",
            unit="$",
        )

        self.feature_deployment_time = self.meter.create_histogram(
            name="manta_feature_deployment_time_days",
            description="Time from feature creation to production deployment",
            unit="d",
            aggregation=ExplicitBucketHistogramAggregation([0.5, 1, 3, 7, 14, 30]),
        )

    def _setup_anomaly_metrics(self) -> None:
        """Setup anomaly detection metrics."""
        self.anomaly_detection_latency = self.meter.create_histogram(
            name="manta_anomaly_detection_latency_ms",
            description="Latency of anomaly detection inference",
            unit="ms",
            aggregation=ExplicitBucketHistogramAggregation([10, 50, 100, 500, 1000]),
        )

        self.anomaly_score = self.meter.create_gauge(
            name="manta_anomaly_score",
            description="Raw anomaly score from Isolation Forest (0-1)",
            unit="1",
        )

        self.pattern_quality_score = self.meter.create_gauge(
            name="manta_pattern_quality_score",
            description="Quality score of detected patterns (0-100)",
            unit="1",
        )

        self.canary_rollout_progress = self.meter.create_gauge(
            name="manta_canary_rollout_progress_percent",
            description="Canary rollout progress percentage (0-100)",
            unit="%",
        )

        self.anomaly_detection_unresolved_count = self.meter.create_gauge(
            name="manta_anomaly_detection_unresolved_count",
            description="Number of unresolved anomalies",
            unit="1",
        )

        self.dbscan_cluster_size = self.meter.create_gauge(
            name="manta_anomaly_dbscan_cluster_size",
            description="Size of largest DBSCAN cluster (indicates drift magnitude)",
            unit="1",
        )

        self.canary_rollout_last_phase_change = self.meter.create_gauge(
            name="manta_canary_rollout_last_phase_change_timestamp",
            description="Unix timestamp of last canary phase change",
            unit="1",
        )

        self.jaeger_span_error_rate = self.meter.create_gauge(
            name="manta_jaeger_span_error_rate",
            description="Jaeger span error rate (0-1)",
            unit="1",
        )


# Singleton instance
_metrics_instance: Optional[MantaMetrics] = None


def get_metrics() -> MantaMetrics:
    """Get or create singleton metrics instance."""
    global _metrics_instance
    if _metrics_instance is None:
        _metrics_instance = MantaMetrics()
    return _metrics_instance
